specs/WebMessageObjects.md

```diff
@@ -0,0 +1,341 @@
+                wil::com_ptr<ICoreWebView2WebMessageReceivedEventArgs>(args)
+                    .query<ICoreWebView2WebMessageReceivedEventArgs2>();
+            wil::com_ptr<ICoreWebView2WebMessageObjectCollectionView>
+                objectsCollection;
+            args2->get_AdditionalObjects(&objectsCollection);
+            unsigned int length;
+            objectsCollection->get_Count(&length);
+            std::vector<std::wstring> paths;
+
+            for (unsigned int i = 0; i < length; i++)
+            {
+                wil::com_ptr<IUnknown> object;
+                objectsCollection->GetValueAtIndex(i, &object);
+
+                wil::com_ptr<ICoreWebView2File> file =
+                    object.query<ICoreWebView2File>();
+                if (file)
+                {
+                    // Add the file to message to be sent back to webview
+                    wil::unique_cotaskmem_string path;
+                    file->get_Path(&path);
+                    paths.push_back(path.get());
+                }
+            }
+
+
+            ProcessFiles(files);
+            return S_OK;
+        })
+        .Get(),
+    &m_webMessageReceivedToken);
+```
+
+```c#
+webView.CoreWebView2.WebMessageReceived += WebView_WebMessageReceivedHandler;
+void WebView_WebMessageReceivedHandler(object sender, CoreWebView2WebMessageReceivedEventArgs args)
+{
+    List<string> paths = new List<string>();
+    foreach (var object in args.AdditionalObjects)
+    {
+        if (object is CoreWebView2File) {
+            paths.Add((CoreWebView2File)(object).path);
+        }
+    }
+}
+```
+
+# API Details
+```c#
+/// Representation of a DOM
+/// (File)[https://developer.mozilla.org/en-US/docs/Web/API/File] object
+/// passed via WebMessage. You can use this object to obtain the path of a
+/// File dropped on WebView2.
+/// \snippet ScenarioDragDrop.cpp DroppedFilePath
+[uuid(f2c19559-6bc1-4583-a757-90021be9afec), object, pointer_default(unique)]
+interface ICoreWebView2File : IUnknown {
+  /// Get the absolute file path.
+  [propget] HRESULT Path([out, retval] LPWSTR* path);
+}
+
+/// Read-only collection of generic objects
+[uuid(b547d2d4-b8d5-4278-a544-c54e66b5f45d), object, pointer_default(unique)]
+interface ICoreWebView2ObjectCollectionView : IUnknown {
+  /// Gets the number of items in the collection.
+  [propget] HRESULT Count([out, retval] UINT32* value);
+
+  /// Gets the object at the specified index. Cast the object to the native type
+  /// to access its specific properties.
+  HRESULT GetValueAtIndex([in] UINT32 index,
+      [out, retval] IUnknown** value);
+}
+
+[uuid(50a798ac-40fa-4634-8fb7-9419e5e8a3f8), object, pointer_default(unique)]
+interface ICoreWebView2WebMessageReceivedEventArgs2 : ICoreWebView2WebMessageReceivedEventArgs {
+  /// Additional received WebMessage objects. To pass `additionalObjects` via
+  /// WebMessage to the app, use the
+  /// `chrome.webview.postMessageWithAdditionalObjects` content API.
```

Comment:
Probably should add a section to the end of Background about our thoughts about naming this method:

This document names the method `chrome.webview.postMessageWithAdditionalObjects`. We considered making it an overload of the existing `chrome.webview.postMessage(message)` but that method is patterned after the existing `MessagePort.postMessage(message, transferList)` DOM method and we're worried about confusion or future compat mixing up the additionalObjects array with the transferList array.